include/cuco/utility/fast_int.cuh

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/__config>
+
+#include <cstdint>
+#include <cuda/std/bit>
+#include <cuda/std/limits>
+#include <cuda/std/type_traits>
+
+namespace cuco::utility {
+
+/**
+ * @brief Integer type with optimized division and modulo operators.
+ *
+ * @tparam T Underlying integer type
+ */
+template <typename T>
+struct fast_int {
+  static_assert(cuda::std::is_same_v<T, std::int32_t> or cuda::std::is_same_v<T, std::uint32_t>
+#if defined(CUCO_HAS_INT128)
+                  or cuda::std::is_same_v<T, std::int64_t> or cuda::std::is_same_v<T, std::uint64_t>
+#endif
+                ,
+                "Unsupported type");
+
+  using value_type = T;  ///< Underlying integer type
+
+  /**
+   * @brief Constructs a fast_int from an integer value.
+   *
+   * @param value Integer value
+   */
+  __host__ __device__ explicit constexpr fast_int(T value) noexcept : value_{value}
+  {
+    evaluate_magic_numbers();
+  }
+
+  /**
+   * @brief Explicit conversion operator to the underlying value type.
+   *
+   * @return Underlying value
+   */
+  __host__ __device__ explicit constexpr operator value_type() const noexcept { return value_; }
+
+ private:
+  using intermediate_type =
+    cuda::std::conditional_t<sizeof(value_type) == 4,
+                             std::uint64_t,
+                             unsigned __int128>;  ///< Intermediate type for multiplication
+  using unsigned_value_type = cuda::std::make_unsigned_t<value_type>;  ///< Unsigned value type
+  using signed_value_type   = cuda::std::make_signed_t<value_type>;    ///< Signed value type
+
+  static constexpr value_type value_bits =
+    CHAR_BIT * sizeof(value_type);  ///< Number of bits required to represent the value
+
+  /**
+   * @brief Computes the high bits of the multiplication of two unsigned integers.
+   *
+   * @param lhs Left-hand side of the multiplication
+   * @param rhs Right-hand side of the multiplication
+   *
+   * @return High bits of the multiplication
+   */
+  __host__ __device__ constexpr value_type mulhi(unsigned_value_type lhs,
+                                                 unsigned_value_type rhs) const noexcept
+  {
+#if defined(__CUDA_ARCH__)
+    if constexpr (sizeof(value_type) == 4) {
+      return __umulhi(lhs, rhs);
+    } else {
+      return __umul64hi(lhs, rhs);
+    }
+#else
+    return (intermediate_type(lhs) * intermediate_type(rhs)) >> value_bits;
+#endif
+  }
+
+  /**
+   * @brief Computes the log2 of an unsigned integer.
+   *
+   * @param v Unsigned integer
+   *
+   * @return Log2 of the unsigned integer
+   */
+  __host__ __device__ constexpr value_type log2(value_type v) const noexcept
+  {
+#if defined(__CUDA_ARCH__)
+    if constexpr (sizeof(value_type) == 4) {
+      return 31 - __clz(signed_value_type(v));
+    } else {
+      return 63 - __clzll(signed_value_type(v));
+    }
+#else
+    return cuda::std::bit_width(unsigned_value_type(v)) - 1;
+#endif
```

Comment:
```suggestion
    return cuda::std::bit_width(unsigned_value_type(v)) - 1;
```
On the device side, ` cuda::std::bit_width` should be as efficient as using `clz` instructions thus the dispatching seems an overkill.